CSP/read_problem.py

```diff
@@ -0,0 +1,71 @@
+# This file will handle the reading in of a problem in 4pw format into our CSP model
+
+from constraint import *
+from find_domain import find_domain
+from rules import generate_constraints
+import sys
+
+
+# Helper function that parses a 3 character harmony string into a len 3 list (e.g. 5Mr -> [5, 'M', 'r'])
+def parse_harmony(harmony_string):
+    return [int(harmony_string[0]) if harmony_string[0].isdigit() else harmony_string[0], harmony_string[1],
+            int(harmony_string[2]) if harmony_string[2].isdigit() else harmony_string[2]]
+
+
+if len(sys.argv) < 2:
+    print('Must provide an input file')
+    exit()
+
+file = sys.argv[1]
+
+# Same scheme as in validate.py
+voices = [[], [], [], []]
+names = ['s', 'a', 't', 'b']  # For variable naming in CSP
+ranges = [(24, 45), (19, 38), (12, 33), (0, 24)]  # Pitch ranges for voices
+harmony = []
+
+with open(file) as input_data:
+    line_index = 0
+
+    for line in input_data:
+        stripped = line.rstrip()
+
+        offset = 2 if line_index < 4 else 3
+
+        for i in range(0, len(stripped), offset):
+            element = stripped[i:i+offset]
+
+            if line_index < 4:
+                voices[line_index].append(int(element) if element.isdigit() else element)
```

Comment:
I do not necessarily like that this plus line 53 forces `??` to be used to designate a missing character.  I would rather anything that is not a number be used to designate that.  Are you okay with `else '??'`?